DPL: make dispatcher lossy

ktf opened the pull request:
This makes the QC dispatcher drop data if the receiving side is not
ready to process them, avoiding building backpressure. Notice
that this also depends on the length of the message queues in FairMQ,
which by default is 4 (which could be already too much) and which
can be user customized.

ktf committed:
DPL: make dispatcher lossy

This makes the QC dispatcher drop data if the receiving side is not
ready to process them, avoiding building backpressure. Notice
that this also depends on the length of the message queues in FairMQ,
which by default is 4 (which could be already too much) and which
can be user customized.

ktf commented:
@knopers8 @Barthelemy, in principle this is all it would be needed now...

knopers8 commented on Framework/Core/src/SendingPolicy.cxx:
```diff
@@ -20,8 +21,12 @@ namespace o2::framework
-    .matcher = [](DeviceSpec const& spec, ConfigContext const& ctx) { return true; },
-    .send = [](FairMQDevice& device, FairMQParts& parts, std::string const& channel) { device.Send(parts, channel, 0); }}};
+            .name = "dispatcher",
+            .matcher = [](DeviceSpec const& spec, ConfigContext const& ctx) { return spec.name == "Dispatcher"; },
```
Instead of this, could you please check if a label "Dispatcher" exists? In the future we might have to add some hashes in the name to allow for multiple Dispatchers.

ktf replied:
I guess something like this?

knopers8 replied:
Yes, thank you

knopers8 reviewed (changes requested):
Thanks, this should be indeed all that we need. Could you consider applying my one comment?

ktf committed:
DPL: add DeviceSpecHelpers::hasLabel and DataProcessorSpecHelpers::hasLabel

ktf commented:
@knopers8 @Barthelemy the batch_test seems to be failing consistently. Could it be misbehaving because of the new lossy setup?

knopers8 commented:
Yes, it appears that the new mechanism is quite lossy indeed. You can see:
```
+ entries=157
+ '[' 157 -ne 200 ']'
```
Now that we allow for losses, this line should be changed, but 25% loss on a CI machine is actually quite worrying. What is the fmq buffer size currently?

ktf commented:
I think the default queue is something like 4 elements, to avoid being able to explode memory. We could either make it longer via a ChannelConfigurationPolicy or try with 1ms timeout.

ktf committed:
Use 1ms timeout for sending objects.

Let's see if this is enough to make sure the receiving side is processing things in the CI.

alibuild committed:
Adapt to new FairLogger API (#104)

knopers8 commented:
Indeed that helps (198/200 sampled received), however I think that in this case the mechanism will not actually allow to unblock dataflow. If message rate is higher than 1000/s, then it will eventually fill the queues.

Is it possible to set queue sizes per device? Maybe we could risk increasing it e.g. to 16 and keeping 0ms timeout as a hopefully good balance.

ktf commented:
Yes, agreed that this only works if the rate is < 1000Hz.

You can customise the size of the queue for a given channel using the ChannelConfigurationPolicy, IIRC. There is something like sndBufSize which you can override...

ktf committed:
Make it customised, but behaving as before.

ktf committed:
Update SendingPolicy.cxx

ktf commented:
Ok, I am changing this to have a separate policy for the dispatcher, but behaving as before. Enabling it will be a matter of turning a -1 in a 0.

ktf commented:
@TimoWilken I see there is now an aliDoctor error every now and then. What is that about?

knopers8 commented:
> Ok, I am changing this to have a separate policy for the dispatcher, but behaving as before. Enabling it will be a matter of turning a -1 in a 0.

Sounds good, thanks

TimoWilken commented:
> @TimoWilken I see there is now an aliDoctor error every now and then. What is that about?

@ktf I think aliDoctor is timing out on hosts where multiple CI builders are running, due to the disk activity from other builders blocking the cleanup steps for the many Docker containers created by aliDoctor. https://github.com/alisw/alibuild/pull/728 should fix this.
